gimmik/kernels/cuda/bstream.mako

```diff
@@ -0,0 +1,56 @@
+<%inherit file='base'/>
+
+__global__ void
+% if n is None:
+${kname}(int n,
+         const ${dtype}* __restrict__ b, int ldb,
+         ${dtype}* __restrict__ c, int ldc)
+{
+  % if width > 1:
+    n = ((n + ${width} - 1) / ${width}) * ${width};
+    ldb /= ${width};
+    ldc /= ${width};
+  % endif
+% else:
+${kname}(const ${dtype}* __restrict__ b, ${dtype}* __restrict__ c)
+{
+    const int n = ${-(-n // width)};
+    const int ldb = ${ldb // width};
+    const int ldc = ${ldc // width};
+% endif
+    const int i = blockDim.x*blockIdx.x + threadIdx.x;
+
+    if (i < n)
+    {
+        ${dtype} bv, csub[${m}];
+
+## Iterare through the used rows of B
+% for kx in bix:
+        bv = __ldcg(b + i + ${kx}*ldb);
+  % for j, jx in enumerate(A[:, kx]):
+    % if jx != 0 and kx == afix[j]:
+        csub[${j}] = ${jx}*bv;
+    % elif jx != 0:
+        csub[${j}] += ${jx}*bv;
+    % endif
+    ##
+    % if kx == alix[j] and beta == 0:
+        __stcg(c + i + ${j}*ldc, csub[${j}]);
+    % elif kx == alix[j] and beta == 1:
+        c[i + ${j}*ldc] += csub[${j}];
+    % elif kx == alix[j]:
+        c[i + ${j}*ldc] = csub[${j}] + ${beta}*c[i + ${j}*ldc];
+    % endif
+  % endfor
+% endfor
+
+## Handle rows of A which are all zero
+% for j, jx in enumerate(afix):
+  % if jx == -1 and beta == 0:
+        c[i + ${j}*ldc] = make_zero();
```

Comment:
Is make zero better than having a macro that does it? Is it guaranteed to get in-lined in all cases?

Reply:
I think a macro would work equally well here.